code/go/0chain.net/smartcontract/dbs/event/challenge.go

```diff
@@ -0,0 +1,108 @@
+package event
+
+import (
+	"errors"
+	"fmt"
+
+	"0chain.net/smartcontract/dbs"
+
+	"0chain.net/core/common"
+	"gorm.io/gorm"
+)
+
+type Challenge struct {
+	gorm.Model
+	ChallengeID    string           `json:"challenge_id" gorm:"index:challenge_id"`
+	CreatedAt      common.Timestamp `json:"created_at" gorm:"created_at"`
+	AllocationID   string           `json:"allocation_id" gorm:"allocation_id"`
+	BlobberID      string           `json:"blobber_id" gorm:"blobber_id"`
+	ValidatorsID   string           `json:"validators_id" gorm:"validators_id"`
+	Seed           int64            `json:"seed" gorm:"seed"`
+	AllocationRoot string           `json:"allocation_root" gorm:"allocation_root"`
+	Responded      bool             `json:"responded" gorm:"responded"`
+}
+
+func (ch *Challenge) exists(edb *EventDb) (bool, error) {
+	err := edb.Get().
+		Model(&Challenge{}).
+		Where(&Challenge{ChallengeID: ch.ChallengeID}).
+		Take(ch).Error
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return false, nil
+		}
+		return false, fmt.Errorf("failed to check challenge existence %v, error %v", ch, err)
+	}
+	return true, nil
+}
+
+func (edb *EventDb) GetChallenge(challengeID string) (*Challenge, error) {
+	var ch Challenge
+
+	result := edb.Store.Get().Model(&Challenge{}).Where(&Challenge{ChallengeID: challengeID}).First(&ch)
+	if result.Error != nil {
+		return nil, fmt.Errorf("error retriving Challenge node with ID %v; error: %v", challengeID, result.Error)
+	}
+
+	return &ch, nil
+}
+
+func (edb *EventDb) GetOpenChallengesForBlobber(blobberID string, now, cct common.Timestamp) ([]*Challenge, error) {
+	var chs []*Challenge
+	expiry := now - cct
+
+	result := edb.Store.Get().Model(&Challenge{}).
+		Where("blobber_id = ? AND responded = ? AND created_at >= ?",
```

Comment:
do we have index for this clause?
what cardinality would be for challenge table

Reply:
> this will be huge table
> as challenges are generated every round